testgrid/cmd/summarizer/summarizer.go

```diff
@@ -0,0 +1,116 @@
+// DashboardTabSummary proto.
+func TableToSummary(queryData Response, dashboardName string, dashboardTab configpb.DashboardTab, updateTimestamp int64) (summarypb.DashboardTabSummary, error) {
+	s := extractBasicSummary(queryData, dashboardName, dashboardTab, updateTimestamp)
+	s.FailingTestSummaries = extractTests(queryData)
+	return s, nil
+}
+
+func extractBasicSummary(queryData Response, dashboardName string, dashboardTab configpb.DashboardTab, updateTimestamp int64) summarypb.DashboardTabSummary {
+	var bugURL string
+	if len(queryData.OpenBugTemplate.URL) > 0 {
+		bugURL = queryData.OpenBugTemplate.URL
+	}
+	lastRunTimestamp := 0
+	for i, e := range queryData.Timestamps {
+		if i == 0 || e > lastRunTimestamp {
+			lastRunTimestamp = e
+		}
+	}
+	overallStatus := summarypb.DashboardTabSummary_UNKNOWN
+	if queryData.OverallStatus > 0 {
+		overallStatus = queryData.OverallStatus
+	}
+
+	s := summarypb.DashboardTabSummary{
+		DashboardName:       dashboardName,
+		DashboardTabName:    dashboardTab.Name,
+		Alert:               queryData.Alerts,
+		LastUpdateTimestamp: float64(updateTimestamp),
+		LastRunTimestamp:    float64(lastRunTimestamp),
+		LatestGreen:         queryData.LatestGreen,
+		OverallStatus:       overallStatus,
+		Status:              queryData.Summary,
+		BugUrl:              bugURL,
+	}
+
+	return s
+}
+
+func appendQuery(query, newKey, newVal string) string {
+	if len(query) > 0 {
+		return fmt.Sprintf("%s&%s=%s", query, newKey, newVal)
+	} else {
+		return fmt.Sprintf("%s=%s", newKey, newVal)
+	}
+}
+
+func extractTests(queryData Response) []*summarypb.FailingTestSummary {
+	var failingTestSummaries []*summarypb.FailingTestSummary
+	if len(queryData.Tests) > 0 {
+		for _, test := range queryData.Tests {
+			target := test.Target
+			failTestID := test.Alert.TestID
+			var failTestLink string
+			if len(failTestID) > 0 {
+				template := queryData.OpenTestTemplate
+				var URLOptions string
+				if _, ok := template.Options["target"]; ok {
+					URLOptions = appendQuery(URLOptions, "target", target)
+				}
+				if _, ok := template.Options["id"]; ok {
+					URLOptions = appendQuery(URLOptions, "id", failTestID)
+				}
+				queryString := url.QueryEscape(URLOptions)
+				failTestLink = fmt.Sprintf("%s?%s", template.URL, queryString)
+			}
+
+			failingTestSummary := summarypb.FailingTestSummary{
+				DisplayName:    test.Name,
+				TestName:       target,
+				FailBuildId:    test.Alert.FailBuildID,
+				FailTimestamp:  float64(test.Alert.FailTime),
+				PassBuildId:    test.Alert.PassBuildID,
+				PassTimestamp:  float64(test.Alert.PassTime),
+				FailCount:      int32(test.Alert.FailCount),
+				BuildLink:      test.Alert.Link,
+				BuildLinkText:  test.Alert.LinkText,
+				BuildUrlText:   test.Alert.URLText,
+				FailureMessage: test.Alert.Message,
+				LinkedBugs:     test.LinkedBugs,
+				FailTestLink:   failTestLink,
+			}
+			failingTestSummaries = append(failingTestSummaries, &failingTestSummary)
```

Comment:
I refactor everything in the for loop above this line out into a `RowToFailingTestSummary(row *Row) TestSummary`-ish function and use that instead